### ISSUE TITLE
Custom metadata are not attached to unhandled errors 

### ISSUE THREAD

**Issue by Olgagr** [bugsnag/bugsnag-js#733]
We noticed that when an unhandled error is reported, the attached metaData is not shown in the report. We attach them like this:

```javascript
// first we configure the bugsnag client
export const bugsnagClient = bugsnag({
  apiKey: BUGSNAG_API_KEY,
  releaseStage: `${ENV}-ng`,
  beforeSend: () => ['development', 'test'].indexOf(ENV) < 0,
  autoCaptureSessions: ['development', 'test'].indexOf(ENV) < 0,
});
```

```javascript
// later in the app when the app is loaded
bugsnagClient.metaData = {
      user: {
        id: currentUser.id,
        name: `${currentUser.firstname} ${currentUser.lastname}`,
        email: currentUser.email,
      },
      company: { id: currentAccount.id, company_name: currentAccount.title },
};
```

When we call deliberately:

```javascript
bugsnagClient.notify(new Error('test'));
```

the metadata are presented in the report. However, any unhandled error doesn't have it. Moreover, when we change the initial configuration of bugsnagClient and set metadata in `beforeSend` callback, again metadata are in the report for unhandled errors.

```javascript
// this works!
export const bugsnagClient = bugsnag({
  apiKey: BUGSNAG_API_KEY,
  releaseStage: `${ENV}-ng`,
  beforeSend: (report) => {
      report.updateMetaData('user', { email: 'test@test.com' });
      return ['development', 'test'].indexOf(ENV) < 0
  },
  autoCaptureSessions: ['development', 'test'].indexOf(ENV) < 0,
});
```

However, this is not an option. At the time we configure the bugsnagClient we don't have metadata we need. How can we modify the metadata for unhandled errors?

Plugin versions we use:
```
@bugsnag/js: 6.5.2
@bugsnag/plugin-angular": 6.5.2
```

**Comment by phillipsam** [comment]
Hi @Olgagr 

It looks like in your configuration where you are setting `beforeSend: () => ['development', 'test'].indexOf(ENV) < 0`, is it this that you are trying to achieve?: https://docs.bugsnag.com/platforms/javascript/configuration-options/#notifyreleasestages 
This setting also applies to `autoCaptureSessions`, so you will not need to manually turn off session capturing for release stages you are not interested in.

Any errors which happen between when you initialise Bugsnag and when you set your metadata will not have any of your custom metadata attached to that error report.

Are you able to try defining a `beforeSend` callback when your app is loaded and you have access to the metadata you need?
Are you able to give an example of how you are triggering your unhandled error?

**Comment by Olgagr** [comment]
@phillipsam Thanks for replay!

 > It looks like in your configuration where you are setting `beforeSend: () => ['development', 'test'].indexOf(ENV) < 0`, is it this that you are trying to achieve?: https://docs.bugsnag.com/platforms/javascript/configuration-options/#notifyreleasestages
> This setting also applies to `autoCaptureSessions`, so you will not need to manually turn off session capturing for release stages you are not interested in.

Ok, thanks for pointing this. The effect is probably the same but of course using `notifyReleaseStages` is more meaningfull.

> 
> Any errors which happen between when you initialise Bugsnag and when you set your metadata will not have any of your custom metadata attached to that error report.

Yeah, this is reasonable. However, we set our metadata as soon as we get user data which is at the app startup. All unhandled errors that happen later should have this metadata but they don't.

> 
> Are you able to try defining a `beforeSend` callback when your app is loaded and you have access to the metadata you need?

Actually, I don't see such possibility. We are using bugsnag in Angular app. According to documentation you need to first create and configure the bugsnagClient:

```
const bugsnagClient = bugsnag({
  apiKey: BUGSNAG_API_KEY,
  releaseStage: `${ENV}-ng`,
  notifyReleaseStages: ['production-ng', 'qa-ng', 'staging-ng'],
  beforeSend: () => {//something}
});
```

Then create an error factory that uses this client:

```
export function errorHandlerFactory() {
  return new BugsnagErrorHandler(bugsnagClient);
}
```

And then pass this factory to app module as a `ErrorHandler` provider:

```
{
    provide: ErrorHandler,
    useFactory: errorHandlerFactory,
}
```

All of this happens before the app starts to get user data. I don't see in the docs that later I can add to bugsnagClient `beforeSend` callback. Or can I?

> Are you able to give an example of how you are triggering your unhandled error?

Well, I'm not able to give an example :) This is why they are unhandled. Bugsnag catches them for me and sends the report.

**Comment by xander-jones** [comment]
Hey @Olgagr, there's a few of options for you to set user data once the app has started. 

Your beforeSend callback can call out to another function to get the data from a store which will have this data once loaded, for example:
```js
var bugsnagClient = bugsnag({
  apiKey: 'your_api_key',
  beforeSend: function (report) {
    var user = yourGetUserFunction()
    report.user = { name: user.name, id: user.id }
  }
})
```

Alternatively, you can use the following code to inject the user data into your config:
```js
bugsnagClient.user = { name: 'the_name', id: 'the_id' }
```

Finally, if you want to do it through a callback, but want to declare this after the initialisation, you can use the following:
```js
bugsnagClient.config.beforeSend.push(
  function (report) {
    report.user = { name: 'the_name', id: 'the_id' };
  }
)
```

Can you try to throw an unhandled exception yourself to see how your application handles it:
```js
throw new Error('Unhandled error')
```
If you're not getting the metadata you're expecting with one of the methods above, please let us know and we'll take a look into why this may be occurring.

**Comment by Olgagr** [comment]
@xander-jones Thanks for answer and help. I've checked two last ways and neither work. However, it seems like this is our app-specific problem. I created a new angular application and set up this in a similar way we have this in the original app. It works. It's really hard to say what is causing problems as this is a big application with a lot of dependencies. 

Thank you for taking the time to help me with this issue! @xander-jones @phillipsam
